### PR TITLE
Update upgrade guide [wip]

### PR DESCRIPTION
**Pull request by ralphtheninja**
I was thinking we could keep this branch alive and tweak the upgrade guide separately as we go. Then just squash it on top of everything, once we're done.

### CODE REVIEW

File: UPGRADING.md [hunk 1]
```diff
@@ -59,12 +59,44 @@ You must now do:
 const testCommon = require('abstract-leveldown/test/common')
 ```
 
-### Default `testCommon` uses unique temporary directories
+### Abstract test suite API changed
 
-This removes the need for cleanup before and/or after tests. As such the `cleanup` method has been removed from `testCommon`. The `lastLocation` method has also been removed as there is no remaining use of it in abstract tests. The `setUp` and `tearDown` methods became noops.
+All `.all()` methods now has the the following signature:
```
Comment (ralphtheninja) [inline UPGRADING.md]: Typo. Fixing.

**Comment by ralphtheninja** [comment]
@vweevers Lets tweak this ^ until we're happy with it and then maybe you could write something about `seek` after that?

**Comment by vweevers** [comment]
> I was thinking we could keep this branch alive and tweak the upgrade guide separately as we go. Then just squash it on top of everything, once we're done.

Nice idea. Avoids getting into merge conflicts every time

File: UPGRADING.md [hunk 2]
```diff
@@ -59,12 +59,44 @@ You must now do:
 const testCommon = require('abstract-leveldown/test/common')
 ```
 
-### Default `testCommon` uses unique temporary directories
+### Abstract test suite API changed
 
-This removes the need for cleanup before and/or after tests. As such the `cleanup` method has been removed from `testCommon`. The `lastLocation` method has also been removed as there is no remaining use of it in abstract tests. The `setUp` and `tearDown` methods became noops.
+All `.all()` methods now has the following signature:
```
Comment (vweevers) [inline UPGRADING.md]: has => have

File: UPGRADING.md [hunk 3]
```diff
@@ -59,12 +59,44 @@ You must now do:
...
+}
+```
+
+This goes for *most* sub tests as well with the exception of `.setUp()` and `.range()` in `test/iterator-range-test.js` which accepts an additional `data` parameter.
+
+### Default `testCommon` was rewritten to be unopinionated
+
+Now `testCommon` only implements two methods, `setUp` and `tearDown`, which became noops.
+
+As part of removing `location`, the abstract tests no longer use `testCommon.location()`. Instead an implementation *must* implement `testCommon.factory()` which should return a fresh database instance. This allows implementations to pass options to their constructor.
```
Comment (vweevers) [inline UPGRADING.md]: Reading this, it isn't exactly clear what "fresh" means.

File: UPGRADING.md [hunk 4]
```diff
@@ -59,12 +59,44 @@ You must now do:
...
+
+The `cleanup` method has been removed from `testCommon` and it's now up to implementations to handle this. The `lastLocation` method has also been removed as there is no remaining use of it in abstract tests.
```
Comment (vweevers) [inline UPGRADING.md]: Missing here is the fact that in most (if not all) cases, cleanup isn't necessary anymore, seeing as `factory` should return a unique db.

File: UPGRADING.md [hunk 5]
```diff
@@ -59,12 +59,44 @@ You must now do:
...
 
 Previously, implementations using the default `testCommon` had to include `rimraf` in their `devDependencies` and browser-based implementations had to exclude `rimraf` from browserify builds. This is no longer the case.
 
+If your implementation is location based we recommend using `tempy` (or similar) to create unique temporary directories. Together with `testCommon.factory()` your setup could now look something like:
```
Comment (vweevers) [inline UPGRADING.md]: Perhaps: "location based" => "disk-based"

File: UPGRADING.md (hunk 4 above)
Comment (ralphtheninja) [reply UPGRADING.md]: I'll reword it.

**Comment by vweevers** [comment]
The "Default testCommon is for disk-based, Node.js implementations only" section can be removed.

**Comment by ralphtheninja** [comment]
> The "Default testCommon is for disk-based, Node.js implementations only" section can be removed.

Aaah right. I was moving stuff around and forgot about it.

File: UPGRADING.md [hunk 6]
```diff
@@ -97,6 +97,10 @@ abstract.all(test, testCommon)
 
 If your implementation tests were calling this directly, simply remove usage.
 
+### Deprecated `snapshot()` test was removed
```
Comment (vweevers) [inline UPGRADING.md]: Actually, it was moved. If you want I'll take this one

Comment (ralphtheninja) [reply UPGRADING.md]: > If you want I'll take this one

Sure!

**Comment by vweevers** [comment]
The diff is a bit ugly because so much changed.

**Comment by ralphtheninja** [comment]
> The diff is a bit ugly because so much changed.

It's fine. I just read the green :) Btw, I can't review since I made the PR. LGTM!

File: UPGRADING.md [hunk 7]
```diff
@@ -134,6 +134,12 @@ If your implementation previously defined the public `iterator.seek(target)`, it
 
 Please see [README.md](README.md) for details.
 
+### Chained batch has been refactored
+
+- The default `_clear` method is no longer a noop; instead it clears the operations queued by `_put` and/or `_del`
+- The `_write` method now takes an `options` object as its first argument
+- The `db` argument in the constructor became mandatory, as well the `_db` property on the instance.
+
```
Comment (ralphtheninja) [inline UPGRADING.md]: LGTM!

**Comment by ralphtheninja** [comment]
LGTM

**Comment by vweevers** [comment]
Will fill the empty sections later.

File: UPGRADING.md [hunk 8]
```diff
@@ -136,6 +136,29 @@ Please see [README.md](README.md) for details.
...
+They return whatever is given. Previously they were opinionated and mostly geared towards string- and Buffer-based storages. Implementations that didn't already define their own serialization should now do so, according to the types that they support. Please refer to the [README](README.md) for recommendations.
+
+### Range options are serialized
+
+Previously, range options like `lt` were passed through as-is, unlike keys.
+
+### The rules for range options have been relaxed
+
+To support them being used as (encoded) lower and upper bounds, `null`, `undefined`, zero-length strings and zero-length buffers became valid as range options. In fact, any type is now valid. This means `db.iterator({ gt: undefined })` is *not* the same as `db.iterator({})`.
```
Comment (vweevers) [inline UPGRADING.md]: can leave out "being used"

File: UPGRADING.md [hunk 9]
```diff
@@ -2,6 +2,13 @@
 
 This document describes breaking changes and how to upgrade. For a complete list of changes including minor and patch releases, please refer to the changelog.
 
+## Table of Contents
+
+- [Unreleased](#unreleased)
+- [v5](#v5)
+- [v4](#v4)
+- [v3](#v3)
```
Comment (ralphtheninja) [inline UPGRADING.md]: :+1:

**Comment by ralphtheninja** [comment]
> @ralphtheninja done, PTAL. I'll rebase on top of master afterwards, because it's becoming painful not being able to browse the code while working on the upgrade guide.

LGTM. Go!

File: UPGRADING.md [hunk 10]
```diff
@@ -32,9 +32,17 @@ function YourDOWN (location) {
 }
 ```
 
+Be sure to include appropriate type checks. If you relied on the default `AbstractLevelDOWN` behavior that would be:
+
+```js
+if (typeof location !== 'string') {
```
Comment (ralphtheninja) [inline UPGRADING.md]: :+1:

**Comment by ralphtheninja** [comment]
@vweevers Maybe squash this on top of master now?

**Pull request merged**